src/environmentd/src/lib.rs

```diff
@@ -296,10 +300,82 @@ pub async fn serve(config: Config) -> Result<Server, anyhow::Error> {
             tracing_handle: config.tracing_handle,
             adapter_client_rx: internal_http_adapter_client_rx,
             active_connection_count: Arc::clone(&active_connection_count),
+            promote_leader: promote_leader_tx,
+            ready_to_promote: ready_to_promote_rx,
         });
         server::serve(internal_http_conns, internal_http_server)
     });
 
+    'leader_promotion: {
+        let Some(deploy_generation) = config.deploy_generation else { break 'leader_promotion };
+        tracing::info!("Requested deploy generation {deploy_generation}");
+        let mut stash = match config
+            .controller
+            .postgres_factory
+            .open_savepoint(config.adapter_stash_url.clone(), tls.clone())
+            .await
+        {
+            Ok(stash) => stash,
+            Err(e) => {
+                if e.can_recover_with_write_mode() {
+                    tracing::info!("Stash doesn't exist so there's no current deploy generation. We won't wait to be leader");
+                    break 'leader_promotion;
+                } else {
+                    return Err(e.into());
+                }
+            }
+        };
+        // TODO: once all stashes have a deploy_generation, don't need to handle the Option
+        let Some(stash_generation) = stash.with_transaction(move |tx| {
+            Box::pin(async move {
+                stash::deploy_generation(&tx).await
+            })
+        }).await? else {
+            tracing::info!("Stash has no generation, not waiting for leader promotion");
```

Comment:
Why does this not wait? Shouldn't it still wait for the promote request?

Reply:
Made it wait. I considered it as a weird case that would only exist while we deploy the feature but having it wait is totally reasonable